### PR TITLE
build: remove transmission-app.js from repo

### LINKED ISSUE #4060: user-submitted changes to transmission-app.js are risky
It's a generated file, so it doesn't make sense to be in the repo.

It's also a minified packed file, so if it _were_ in the repo it's impossible to review PRs that update it. See https://github.com/transmission/transmission/issues/4024 which recently introduced debug code by accident. PRs that update `transmission-app.js` are a potential attack vector.

### PR DESCRIPTION
**Pull request by ckerr**
Fixes #4060.

Related issue: #4042

### CODE REVIEW

**Comment by Coeur** [comment]
Let's note for context that CPack 3.19 is from 2 years ago:
https://www.kitware.com/cmake-3-19-0-available-for-download/

**Comment by nevack** [comment]
> Let's note for context that CPack 3.19 is from 2 years ago: [kitware.com/cmake-3-19-0-available-for-download](https://www.kitware.com/cmake-3-19-0-available-for-download/) (and we're at 3.24 now, very soon 3.25)

There're some "legacy" systems we need to support, btw

**Comment by GaryElshaw** [comment]
True, but i don't think cpack, in particular, would be an issue on those systems.

**Comment by ckerr** [comment]
> True, but i don't think cpack, in particular, would be an issue on those systems.

Easy enough to test by making a PR that changes the first line in the top-level CMakeLists PR from 3.9 to 3.19 and see if any of the CI boxes scream :scream:

**Comment by mikedld** [comment]
I'm not concerned about p.3 and p.4 as I think people rarely expect to be able to develop native and web clients together. The only functionality in native clients that matters is that they're able to open a web browser with a correct location; whether web client functions properly is then a separate thing to debug, and it's seemingly okay to use some other, more natural [than CMake] means to (re)build it if/when necessary.

For p.5 ("What about people building from git and then running make install without using a tarball?"), this is essentially what happens when we build an installer package (.msi on Windows, and hopefully .dmg on Mac in the future), not something specific to people building from git. We can certainly adjust CI configuration to perform those builds off the tarball, but it's an additional step that's probably better be avoided given that we already have `ENABLE_WEB` option (I'm not sure how well it works right now, especially in conjunction with `make install`, and it also doesn't support `AUTO` which might be nice).

**Comment by ckerr** [comment]
> I'm not concerned about p.3 and p.4 as I think people rarely expect to be able to develop native and web clients together. 

What I mean is, I'm not referring to development here, I'm thinking of end user installs -- don't we want to bundle `transmission-app.js` with the macOS and Windows installs so people can remote into their home PCs from somewhere else? And if so, don't we need to have the scripts that build those releases also build `transmission-app.js`?

**Comment by mikedld** [comment]
Then I guess your answer is in the second paragraph, i.e. yes, we do bundle it right now and yes, we do want it to work.

**Comment by mikedld** [comment]
Thinking about it some more, prebuilt web client should probably be a separate build artifact. It's not a source, so why would we put it into a source tarball?..

**Comment by ckerr** [comment]
> Thinking about it some more, prebuilt web client should probably be a separate build artifact. It's not a source, so why would we put it into a source tarball?..

Think of it in the same category as an `transmission.png` icon file or `transmission.desktop` desktop file for the GTK or Qt clients. It's something that needs to be bundled along in the macOS, GTK, and Qt builds so that they can be driven via remote control.

**Comment by Coeur** [comment]
As it is (and even accounting for Mike's branch **build/remove-transmission-app.js-from-repo-tmp**, this would break support for web when building from Xcode instead of cmake.

So I invite you @ckerr to change the milestone to Transmission 5.0.0.

File: docs/Building-Transmission.md [hunk 1]
```diff
@@ -5,6 +5,8 @@ The source code for both official and nightly releases can be found on our [down
 Software prerequisites:
  * macOS 10.14.4 or newer
  * Xcode 11.3.1 or newer
+ * Node.js 10 or newer
+ * Yarn
```
Comment (Coeur) [inline docs/Building-Transmission.md:9]: Do we still need Yarn after #4637 ?

**Review by Coeur — CHANGES_REQUESTED** [review]
As it is, this PR would break support for web when building from Xcode instead of cmake.

So this needs the addition of a run script in Xcode that will regenerate transmission-app.js (similar to what we do with update-version-h.sh).

File: docs/Building-Transmission.md (hunk 1 above)
Comment (ckerr) [reply docs/Building-Transmission.md:9]: No. It's 2am here so I'm not pushing any more commits tonight, but I'll leave this convo unresolved to remind me when I come back to this

**Comment by ckerr** [comment]
OK looks like there are two remaining tasks

- @mikedld please install Node >= 10 + npm on the TC machines (sanity, sonarcloud)

- I need to add an xcode glue script to this PR as @Coeur pointed out

Once those two tasks are done I think #4060 will be fixed.

**Comment by ckerr** [comment]
> @ckerr, I'm not sure I understand why web client is now a hard requirement. I can certainly look into installing NodeJS but somehow it doesn't feel right. There're many valid use-cases where our web client is just not necessary.

All of the GUI apps can serve the web client & RPC, so when would we want to leave transmission-app.js out? e.g. if someone turns on remote access to their transmission-mac session, they won't be able to connect.

IMO it's either this or keep the generated file in the repo like we're doing now, and regenerate it manually every time we release :shrug:, or some other alternative I haven't thought of.

I do think this adds more complexity to the build, and would be happy for a better alternative.

**Comment by ckerr** [comment]
@mikedld what is the alternative then; how do we remove the generated file from the repo _and_ not build it _and_ bundle it in our releases ...

**Comment by Coeur** [comment]
There should still be an Xcode script of the style:
> "If node/npm is missing, print a warning, otherwise build transmission-app.js"

So that those building the project by themselves get no regressions.

**Comment by ckerr** [comment]
I might have another idea

**Comment by ckerr** [comment]
Superseded by #4658.

**Pull request closed without merging**